File: go/indicators/welleswilder/directionalindicatorplus/directionalindicatorplus.go
package directionalindicatorplus

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/welleswilder/averagetruerange"
	"zpano/indicators/welleswilder/directionalmovementplus"
)

const (
	dipMnemonic    = "+di"
	dipDescription = "Directional Indicator Plus"
	epsilon        = 1e-8
)

// DirectionalIndicatorPlus is Welles Wilder's Directional Indicator Plus (+DI).
//
// The directional indicator plus measures the percentage of the average true range
// that is attributable to upward movement. It is calculated as:
//
//	+DI = 100 * +DM(n) / (ATR * length)
//
// where +DM(n) is the Wilder-smoothed directional movement plus and ATR is the
// average true range over the same length.
//
// The indicator requires close, high, and low values. ATR uses all three;
// directional movement plus uses high and low.
//
// Reference:
//
// Wilder, J. Welles Jr. (1978). New Concepts in Technical Trading Systems.
type DirectionalIndicatorPlus struct {
	mu                      sync.RWMutex
	length                  int
	value                   float64
	averageTrueRange        *averagetruerange.AverageTrueRange
	directionalMovementPlus *directionalmovementplus.DirectionalMovementPlus
}

// NewDirectionalIndicatorPlus returns a new instance of the Directional Indicator Plus indicator.
func NewDirectionalIndicatorPlus(length int) (*DirectionalIndicatorPlus, error) {
	if length < 1 {
		return nil, fmt.Errorf("invalid length %d: must be >= 1", length)
	}

	atr, err := averagetruerange.NewAverageTrueRange(length)
	if err != nil {
		return nil, fmt.Errorf("failed to create average true range: %w", err)
	}

	dmp, err := directionalmovementplus.NewDirectionalMovementPlus(length)
	if err != nil {
		return nil, fmt.Errorf("failed to create directional movement plus: %w", err)
	}

	return &DirectionalIndicatorPlus{
		length:                  length,
		value:                   math.NaN(),
		averageTrueRange:        atr,
		directionalMovementPlus: dmp,
	}, nil
}

// Length returns the length parameter.
func (s *DirectionalIndicatorPlus) Length() int {
	return s.length
}

// IsPrimed indicates whether the indicator is primed.
func (s *DirectionalIndicatorPlus) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.averageTrueRange.IsPrimed() && s.directionalMovementPlus.IsPrimed()
}

// Metadata describes the output data of the indicator.
func (s *DirectionalIndicatorPlus) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.DirectionalIndicatorPlus,
		dipMnemonic,
		dipDescription,
		[]core.OutputText{
			{Mnemonic: dipMnemonic, Description: dipDescription},
			{Mnemonic: "+dm", Description: "Directional Movement Plus"},
			{Mnemonic: "atr", Description: "Average True Range"},
			{Mnemonic: "tr", Description: "True Range"},
		},
	)
}

// Update updates the Directional Indicator Plus given the next bar's close, high, and low values.
func (s *DirectionalIndicatorPlus) Update(close, high, low float64) float64 {
	if math.IsNaN(close) || math.IsNaN(high) || math.IsNaN(low) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	atrValue := s.averageTrueRange.Update(close, high, low)
	dmpValue := s.directionalMovementPlus.Update(high, low)

	if s.averageTrueRange.IsPrimed() && s.directionalMovementPlus.IsPrimed() {
		atrScaled := atrValue * float64(s.length)

		if math.Abs(atrScaled) < epsilon {
			s.value = 0
		} else {
			s.value = 100 * dmpValue / atrScaled //nolint:mnd
		}

		return s.value
	}

	return math.NaN()
}

// UpdateSample updates the Directional Indicator Plus using a single sample value
// as a substitute for close, high, and low.
func (s *DirectionalIndicatorPlus) UpdateSample(sample float64) float64 {
	return s.Update(sample, sample, sample)
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *DirectionalIndicatorPlus) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (s *DirectionalIndicatorPlus) UpdateBar(sample *entities.Bar) core.Output {
	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(sample.Close, sample.High, sample.Low)}

	return output
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *DirectionalIndicatorPlus) UpdateQuote(sample *entities.Quote) core.Output {
	v := (sample.Bid + sample.Ask) / 2 //nolint:mnd

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *DirectionalIndicatorPlus) UpdateTrade(sample *entities.Trade) core.Output {
	v := sample.Price

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: s.Update(v, v, v)}

	return output
}

File: ts/indicators/welles-wilder/directional-indicator-plus/directional-indicator-plus.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { AverageTrueRange } from '../average-true-range/average-true-range';
import { DirectionalMovementPlus } from '../directional-movement-plus/directional-movement-plus';

const dipMnemonic = '+di';
const dipDescription = 'Directional Indicator Plus';
const epsilon = 1e-8;

/**
 * Welles Wilder's Directional Indicator Plus (+DI).
 *
 * The directional indicator plus measures the percentage of the average true range
 * that is attributable to upward movement. It is calculated as:
 *
 *   +DI = 100 * +DM(n) / (ATR * length)
 *
 * where +DM(n) is the Wilder-smoothed directional movement plus and ATR is the
 * average true range over the same length.
 *
 * The indicator requires close, high, and low values.
 */
export class DirectionalIndicatorPlus implements Indicator {

  private readonly length_: number;
  private value_ = NaN;
  private readonly averageTrueRange: AverageTrueRange;
  private readonly directionalMovementPlus: DirectionalMovementPlus;

  constructor(length: number) {
    if (length < 1) {
      throw new Error(`invalid length ${length}: must be >= 1`);
    }

    this.length_ = length;
    this.averageTrueRange = new AverageTrueRange(length);
    this.directionalMovementPlus = new DirectionalMovementPlus(length);
  }

  /** Returns the length parameter. */
  public get length(): number {
    return this.length_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.averageTrueRange.isPrimed() && this.directionalMovementPlus.isPrimed();
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.DirectionalIndicatorPlus,
      dipMnemonic,
      dipDescription,
      [
        { mnemonic: dipMnemonic, description: dipDescription },
        { mnemonic: '+dm', description: 'Directional Movement Plus' },
        { mnemonic: 'atr', description: 'Average True Range' },
        { mnemonic: 'tr', description: 'True Range' },
      ],
    );
  }

  /** Updates the Directional Indicator Plus given the next bar's close, high, and low values. */
  public update(close: number, high: number, low: number): number {
    if (isNaN(close) || isNaN(high) || isNaN(low)) {
      return NaN;
    }

    const atrValue = this.averageTrueRange.update(close, high, low);
    const dmpValue = this.directionalMovementPlus.update(high, low);

    if (this.averageTrueRange.isPrimed() && this.directionalMovementPlus.isPrimed()) {
      const atrScaled = atrValue * this.length_;

      if (Math.abs(atrScaled) < epsilon) {
        this.value_ = 0;
      } else {
        this.value_ = 100 * dmpValue / atrScaled;
      }

      return this.value_;
    }

    return NaN;
  }

  /** Updates the Directional Indicator Plus using a single sample value as a substitute for close, high, and low. */
  public updateSample(sample: number): number {
    return this.update(sample, sample, sample);
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = sample.value;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(sample.close, sample.high, sample.low);
    return [scalar];
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = (sample.bidPrice + sample.askPrice) / 2;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const v = sample.price;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = this.update(v, v, v);
    return [scalar];
  }
}
